### DISCUSSION TITLE
Run getServerSideProps before getInitialProps

### DISCUSSION THREAD

**Discussion by twavv** [vercel/next.js#18177] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

I have an existing `getInitialProps` function defined in `_app`. It sets up Apollo and does a tree render.

I also want to use `getServerSideProps` in a page, but am running into issues because `getInitialProps` runs before `getServerSideProps` and the component assumes that the props provided by `getServerSideProps` is available (it loads content from my CMS).

## Describe the solution you'd like

Run `getServerSideProps` first so that `NextApp.getInitialProps` includes those values.

## Describe alternatives you've considered

The easy thing is to say "don't use tree rendering" but I would rather not have to migrate the whole app for that.

**Comment by twavv** [comment]
I'm not sure why exactly this was converted from an issue (it was confusing - GitHub doesn't notify you that this has happened, and I couldn't find the original issue I posted - I had to go back to my browser history to find it to make sure that I didn't dream posting it).

Also seems unfortunate since discussions don't get nearly the same amount of attention as issues (and if this doesn't count as a feature request - what does? why does that GitHub issue template exist)?

At any rate, I was able to come up with a VERY HACKY way of doing this for my own purposes, though I'd really love for it to be solved in Next itself.

tl;dr: It's an API inspired by Suspense. Throw a specific exception if you detect your in an SSR and don't have the stuff from `getServerSideProps` and have `_app.getInitialProps` catch it, invoke the async function, and re-render. This requires doing a bunch of stuff (eg, I have a webpack plugin that defines `__SERVER__`, which is only true for server-side builds), and you have to deal with the fact that `getServerSideProps` might be invoked more than once (I use a `WeakMap` whose keys are the `http.IncomingMessage` objects).

I'm simultaneously proud and horrified at this solution. I hope it can be fixed by Next (I can't think of scenarios where you'd want getInitialProps to run before getServerSideProps considering that the latter **blocks page transitions** on client-side renders).
